File: test/nest/messages/think_tags_test.exs
defmodule Nest.Messages.ThinkTagsTest do
  @moduledoc """
  Tests for `Nest.Messages.ThinkTags`.

  Covers:
  - No markers → returns input unchanged.
  - Single block → removed.
  - Multiple blocks → all removed.
  - Nested `<think>` inside a think block → outer block
    consumes the matching close; inner is text inside.
  - Empty think block (`<think></think>`) → removed.
  - Orphan `</think>` (no matching `<think>`) → orphan and
    everything after is removed.
  - Orphan `<think>` (no matching `</think>`) → from orphan
    to end is removed.
  - Non-binary input → returns "".
  - `strip_from_parts/1` rewrites `Part.Text`, passes other
    kinds through unchanged.
  - Stray `</think>\n\n` (the OpenAI-style regression) → the
    stray text and tail are removed.
  """
  use ExUnit.Case, async: true

  alias Nest.Messages.Part
  alias Nest.Messages.ThinkTags

  describe "strip/1 — no markers" do
    test "returns the input unchanged" do
      assert ThinkTags.strip("hello world") == "hello world"
    end

    test "returns empty string unchanged" do
      assert ThinkTags.strip("") == ""
    end

    test "returns non-binary input as empty string" do
      assert ThinkTags.strip(nil) == ""
      assert ThinkTags.strip(42) == ""
      assert ThinkTags.strip(:atom) == ""
      assert ThinkTags.strip(%{}) == ""
    end
  end

  describe "strip/1 — single block" do
    test "removes a single <think>...</think> block" do
      assert ThinkTags.strip("before<think>reasoning</think>after") ==
               "beforeafter"
    end

    test "removes a block that takes the whole string" do
      assert ThinkTags.strip("<think>reasoning</think>") == ""
    end

    test "removes a block with surrounding newlines" do
      assert ThinkTags.strip("line1\n<think>\nreasoning\n</think>\nline2\n") ==
               "line1\n\nline2\n"
    end
  end

  describe "strip/1 — multiple blocks" do
    test "removes all blocks" do
      assert ThinkTags.strip("a<think>1</think>b<think>2</think>c") == "abc"
    end

    test "preserves text between blocks" do
      assert ThinkTags.strip("<think>a</think> middle <think>b</think>") == " middle "
    end
  end

  describe "strip/1 — nested and edge cases" do
    test "treats nested <think> inside a think block as text" do
      # The inner `<think>...end` is text inside the outer
      # think block. The outer block closes at the FINAL
      # `</think>`. The inner `</think>` (after `inner`) is
      # consumed to keep the outer block from closing early.
      assert ThinkTags.strip("a<think>outer<think>inner</think>end</think>b") ==
               "ab"
    end

    test "removes an empty think block" do
      assert ThinkTags.strip("before<think></think>after") == "beforeafter"
    end

    test "removes an orphan </think> (no matching <think>)" do
      # Stray closing tag and everything after is reasoning.
      assert ThinkTags.strip("before</think>after") == "before"
    end

    test "removes an orphan <think> (no matching </think>)" do
      # Incomplete think block; the rest is reasoning.
      assert ThinkTags.strip("before<think>reasoning") == "before"
    end

    test "treats </think>\\n\\n as an orphan closing (OpenAI-style regression)" do
      # The exact stray-text pattern from the regression:
      # `</think>\n\n` leaking into the assistant's reply.
      assert ThinkTags.strip("hello</think>\n\nworld") == "hello"
    end
  end

  describe "strip_from_parts/1" do
    test "rewrites Part.Text" do
      parts = [
        %Part.Text{text: "before<think>reasoning</think>after"},
        %Part.Text{text: "no markers"}
      ]

      assert ThinkTags.strip_from_parts(parts) == [
               %Part.Text{text: "beforeafter"},
               %Part.Text{text: "no markers"}
             ]
    end

    test "passes other part kinds through unchanged" do
      tool_use = %Part.ToolUse{id: "1", name: "x", arguments: %{}}
      tool_result = %Part.ToolResult{tool_call_id: "1", name: "x", content: "ok"}
      thinking = %Part.Thinking{thinking: "inner reasoning"}
      refusal = %Part.Refusal{refusal: "no"}

      parts = [
        tool_use,
        %Part.Text{text: "<think>r</think>after"},
        tool_result,
        thinking,
        refusal
      ]

      assert ThinkTags.strip_from_parts(parts) == [
               tool_use,
               %Part.Text{text: "after"},
               tool_result,
               thinking,
               refusal
             ]
    end

    test "returns empty list for non-list input" do
      assert ThinkTags.strip_from_parts(nil) == []
      assert ThinkTags.strip_from_parts(42) == []
      assert ThinkTags.strip_from_parts(:atom) == []
    end

    test "returns empty list for empty list" do
      assert ThinkTags.strip_from_parts([]) == []
    end
  end
end

File: assets/js/utils/thinkTags.test.js
import { describe, expect, it } from "vitest";
import { splitThinkTags, splitThinkFromParts } from "./thinkTags.js";

describe("splitThinkTags", () => {
  it("returns no segments for empty / non-string input", () => {
    expect(splitThinkTags("")).toEqual([]);
    expect(splitThinkTags(null)).toEqual([]);
    expect(splitThinkTags(undefined)).toEqual([]);
    expect(splitThinkTags(42)).toEqual([]);
  });

  it("returns a single text segment when there are no markers", () => {
    expect(splitThinkTags("hello world")).toEqual([
      { kind: "text", text: "hello world" },
    ]);
  });

  it("splits a single <think>...</think> block", () => {
    expect(splitThinkTags("before<think>reasoning</think>after")).toEqual([
      { kind: "text", text: "before" },
      { kind: "thinking", text: "reasoning" },
      { kind: "text", text: "after" },
    ]);
  });

  it("splits multiple blocks in the same string", () => {
    expect(splitThinkTags("a<think>1</think>b<think>2</think>c")).toEqual([
      { kind: "text", text: "a" },
      { kind: "thinking", text: "1" },
      { kind: "text", text: "b" },
      { kind: "thinking", text: "2" },
      { kind: "text", text: "c" },
    ]);
  });

  it("handles an empty think block (<think></think>)", () => {
    expect(splitThinkTags("before<think></think>after")).toEqual([
      { kind: "text", text: "before" },
      { kind: "thinking", text: "" },
      { kind: "text", text: "after" },
    ]);
  });

  it("handles an orphan closing </think> (no matching <think>)", () => {
    // Stray `</think>` in the response is treated as a thinking
    // marker — the orphan text and everything after is routed
    // to the thinking channel. The orphan itself is consumed
    // (so the user doesn't see raw `</think>` characters).
    expect(splitThinkTags("before</think>after")).toEqual([
      { kind: "text", text: "before" },
      { kind: "thinking", text: "after" },
    ]);
  });

  it("handles an orphan <think> (no matching </think>)", () => {
    // The open `<think>` is consumed and everything from there
    // to the end of the string is routed to the thinking
    // channel (an incomplete think block).
    expect(splitThinkTags("before<think>reasoning")).toEqual([
      { kind: "text", text: "before" },
      { kind: "thinking", text: "reasoning" },
    ]);
  });

  it("does not recurse into a nested <think> inside a think block", () => {
    // The inner `<think>` is just text inside the outer think
    // block — we don't open a new thinking segment.
    expect(
      splitThinkTags("a<think>outer<think>inner</think>end</think>b"),
    ).toEqual([
      { kind: "text", text: "a" },
      { kind: "thinking", text: "outer<think>inner</think>end" },
      { kind: "text", text: "b" },
    ]);
  });

  it("preserves whitespace and newlines in segments", () => {
    expect(
      splitThinkTags("line1\n<think>\nreasoning\n\nmore\n</think>\nline2\n"),
    ).toEqual([
      { kind: "text", text: "line1\n" },
      { kind: "thinking", text: "\nreasoning\n\nmore\n" },
      { kind: "text", text: "\nline2\n" },
    ]);
  });

  it("treats </think>\\n\\n as thinking (orphan closing)", () => {
    // The exact stray-text pattern from the regression:
    // </think>\n\n leaking into the assistant's reply.
    expect(splitThinkTags("hello</think>\n\nworld")).toEqual([
      { kind: "text", text: "hello" },
      { kind: "thinking", text: "\n\nworld" },
    ]);
  });
});

describe("splitThinkFromParts", () => {
  it("returns null/empty for non-array input", () => {
    expect(splitThinkFromParts(null)).toEqual({
      thinking: null,
      textParts: [],
    });
    expect(splitThinkFromParts(undefined)).toEqual({
      thinking: null,
      textParts: [],
    });
  });

  it("returns the parts unchanged when no think markers are present", () => {
    const parts = [
      { kind: "thinking", thinking: "reasoning" },
      { kind: "text", text: "hello" },
    ];
    expect(splitThinkFromParts(parts)).toEqual({
      thinking: "reasoning",
      textParts: [{ kind: "text", text: "hello" }],
    });
  });

  it("concatenates multiple Part.Thinking entries with a blank-line separator", () => {
    const parts = [
      { kind: "thinking", thinking: "first" },
      { kind: "thinking", thinking: "second" },
    ];
    expect(splitThinkFromParts(parts)).toEqual({
      thinking: "first\n\nsecond",
      textParts: [],
    });
  });

  it("splits a <think> block inside a Part.Text", () => {
    const parts = [
      { kind: "text", text: "before<think>reasoning</think>after" },
    ];
    const result = splitThinkFromParts(parts);
    expect(result.thinking).toBe("reasoning");
    expect(result.textParts).toEqual([
      { kind: "text", text: "before" },
      { kind: "text", text: "after" },
    ]);
  });

  it("combines a Part.Thinking and a <think> in Part.Text", () => {
    const parts = [
      { kind: "thinking", thinking: "alpha" },
      { kind: "text", text: "before<think>beta</think>after" },
    ];
    const result = splitThinkFromParts(parts);
    expect(result.thinking).toBe("alpha\n\nbeta");
    expect(result.textParts).toEqual([
      { kind: "text", text: "before" },
      { kind: "text", text: "after" },
    ]);
  });

  it("preserves non-text/non-thinking parts unchanged", () => {
    const toolUse = { kind: "tool_use", id: "1", name: "x", arguments: {} };
    const toolResult = {
      kind: "tool_result",
      toolCallId: "1",
      content: "ok",
    };
    const parts = [
      { kind: "text", text: "before" },
      toolUse,
      { kind: "text", text: "<think>r</think>after" },
      toolResult,
    ];
    const result = splitThinkFromParts(parts);
    expect(result.thinking).toBe("r");
    expect(result.textParts).toEqual([
      { kind: "text", text: "before" },
      toolUse,
      { kind: "text", text: "after" },
      toolResult,
    ]);
  });

  it("returns null thinking when no think content is present", () => {
    const parts = [
      { kind: "text", text: "hello" },
      { kind: "text", text: "world" },
    ];
    const result = splitThinkFromParts(parts);
    expect(result.thinking).toBeNull();
    expect(result.textParts).toEqual([
      { kind: "text", text: "hello" },
      { kind: "text", text: "world" },
    ]);
  });

  it("skips null/undefined entries in the parts array", () => {
    const parts = [
      null,
      { kind: "text", text: "before<think>r</think>after" },
      undefined,
    ];
    const result = splitThinkFromParts(parts);
    expect(result.thinking).toBe("r");
    expect(result.textParts).toEqual([
      { kind: "text", text: "before" },
      { kind: "text", text: "after" },
    ]);
  });
});
